File: yinwei/contracts/v1/js/scene_contract_v1.mjs
//! SceneContractV1 — authoritative scene document validation.

const SCENE_SCHEMA_VERSION_V1 = 1;
const MAX_SOURCES_V1 = 32;
const MAX_EMITTERS_V1 = 8;
const MAX_ID_LENGTH_V1 = 64;
const QUAT_DEGENERATE_NORM_V1 = 1e-12;

export const FORBIDDEN_SCENE_TOP_LEVEL_KEYS_V1 = [
  'camera',
  'selection',
  'hover',
  'drag',
  'telemetry',
  'audio',
  'spatialParams',
  'arrayLayout',
];

const ID_PATTERN = /^[A-Za-z0-9_\-:.]+$/;

export function decideRevisionV1(appliedRevision, incomingRevision) {
  if (incomingRevision <= appliedRevision) {
    return { apply: false, reason: 'stale_revision', discontinuity: false };
  }
  return {
    apply: true,
    discontinuity: incomingRevision > appliedRevision + 1,
  };
}

function isFiniteNumber(v) {
  return typeof v === 'number' && Number.isFinite(v);
}

function validId(id) {
  return typeof id === 'string' && id.length > 0 && id.length <= MAX_ID_LENGTH_V1 && ID_PATTERN.test(id);
}

function validateObject(obj, expectedType) {
  if (!obj || typeof obj !== 'object') return 'invalid_object';
  if (!validId(obj.id)) return 'invalid_id';
  if (obj.type !== expectedType) return 'wrong_type';
  const pos = obj.worldPosition;
  if (!pos || typeof pos !== 'object') return 'non_finite_position';
  if (!isFiniteNumber(pos.x) || !isFiniteNumber(pos.y) || !isFiniteNumber(pos.z)) {
    return 'non_finite_position';
  }
  if (typeof obj.enabled !== 'boolean' || typeof obj.active !== 'boolean') {
    return 'invalid_flags';
  }
  if (Object.prototype.hasOwnProperty.call(obj, 'orientation')) {
    const q = obj.orientation;
    if (!q || typeof q !== 'object') return 'invalid_orientation';
    if (!isFiniteNumber(q.w) || !isFiniteNumber(q.x) || !isFiniteNumber(q.y) || !isFiniteNumber(q.z)) {
      return 'invalid_orientation';
    }
    const n2 = q.w * q.w + q.x * q.x + q.y * q.y + q.z * q.z;
    if (n2 < QUAT_DEGENERATE_NORM_V1 * QUAT_DEGENERATE_NORM_V1) {
      return 'invalid_orientation';
    }
  }
  return null;
}

export function validateSceneV1(scene) {
  if (!scene || typeof scene !== 'object') {
    return { valid: false, reason: 'invalid_document' };
  }
  for (const key of FORBIDDEN_SCENE_TOP_LEVEL_KEYS_V1) {
    if (Object.prototype.hasOwnProperty.call(scene, key)) {
      return { valid: false, reason: 'forbidden_key' };
    }
  }
  if (scene.schemaVersion !== SCENE_SCHEMA_VERSION_V1) {
    return { valid: false, reason: 'schema_version' };
  }
  if (!Number.isInteger(scene.revision) || scene.revision < 0) {
    return { valid: false, reason: 'invalid_revision' };
  }
  const listenerErr = validateObject(scene.listener, 'listener');
  if (listenerErr) return { valid: false, reason: listenerErr };
  if (!Array.isArray(scene.sources) || !Array.isArray(scene.emitters)) {
    return { valid: false, reason: 'invalid_collections' };
  }
  if (scene.sources.length > MAX_SOURCES_V1) {
    return { valid: false, reason: 'too_many_sources' };
  }
  if (scene.emitters.length > MAX_EMITTERS_V1) {
    return { valid: false, reason: 'too_many_emitters' };
  }
  const ids = new Set();
  ids.add(scene.listener.id);
  for (const item of scene.sources) {
    const err = validateObject(item, 'source');
    if (err) return { valid: false, reason: err };
    if (ids.has(item.id)) return { valid: false, reason: 'duplicate_id' };
    ids.add(item.id);
  }
  for (const item of scene.emitters) {
    const err = validateObject(item, 'emitter');
    if (err) return { valid: false, reason: err };
    if (ids.has(item.id)) return { valid: false, reason: 'duplicate_id' };
    ids.add(item.id);
  }
  return { valid: true, reason: null };
}

File: yinwei/apps/yinwei_player/lib/contracts/scene_contract_v1.dart
/// SceneContractV1 — authoritative Spatial Scene Domain document.
///
/// Isolated from audio configuration, UI selection, camera, and telemetry.
library;

const int kSceneSchemaVersionV1 = 1;
const int kMaxSourcesV1 = 32;
const int kMaxEmittersV1 = 8;
const int kMaxIdLengthV1 = 64;
const double kQuatDegenerateNormV1 = 1e-12;

const List<String> kForbiddenSceneTopLevelKeysV1 = [
  'camera',
  'selection',
  'hover',
  'drag',
  'telemetry',
  'audio',
  'spatialParams',
  'arrayLayout',
];

final _idPattern = RegExp(r'^[A-Za-z0-9_\-:.]+$');

class SceneValidationV1 {
  const SceneValidationV1.valid()
      : valid = true,
        reason = null;

  const SceneValidationV1.invalid(this.reason) : valid = false;

  final bool valid;
  final String? reason;
}

enum RevisionDecisionV1 { apply, rejectStale }

class RevisionVerdictV1 {
  const RevisionVerdictV1({
    required this.decision,
    required this.discontinuity,
  });

  final RevisionDecisionV1 decision;
  final bool discontinuity;

  bool get apply => decision == RevisionDecisionV1.apply;
}

RevisionVerdictV1 decideRevisionV1({
  required int appliedRevision,
  required int incomingRevision,
}) {
  if (incomingRevision <= appliedRevision) {
    return const RevisionVerdictV1(
      decision: RevisionDecisionV1.rejectStale,
      discontinuity: false,
    );
  }
  return RevisionVerdictV1(
    decision: RevisionDecisionV1.apply,
    discontinuity: incomingRevision > appliedRevision + 1,
  );
}

bool _isFiniteNum(Object? v) {
  if (v is! num) return false;
  return v.isFinite;
}

bool _validId(Object? id) {
  if (id is! String) return false;
  if (id.isEmpty || id.length > kMaxIdLengthV1) return false;
  return _idPattern.hasMatch(id);
}

String? _validateObject(
  Map<dynamic, dynamic> obj, {
  required String expectedType,
}) {
  if (!_validId(obj['id'])) return 'invalid_id';
  if (obj['type'] != expectedType) return 'wrong_type';
  final pos = obj['worldPosition'];
  if (pos is! Map) return 'non_finite_position';
  if (!_isFiniteNum(pos['x']) ||
      !_isFiniteNum(pos['y']) ||
      !_isFiniteNum(pos['z'])) {
    return 'non_finite_position';
  }
  if (obj['enabled'] is! bool || obj['active'] is! bool) {
    return 'invalid_flags';
  }
  if (obj.containsKey('orientation')) {
    final q = obj['orientation'];
    if (q is! Map) return 'invalid_orientation';
    if (!_isFiniteNum(q['w']) ||
        !_isFiniteNum(q['x']) ||
        !_isFiniteNum(q['y']) ||
        !_isFiniteNum(q['z'])) {
      return 'invalid_orientation';
    }
    final w = (q['w'] as num).toDouble();
    final x = (q['x'] as num).toDouble();
    final y = (q['y'] as num).toDouble();
    final z = (q['z'] as num).toDouble();
    final n = w * w + x * x + y * y + z * z;
    if (n < kQuatDegenerateNormV1 * kQuatDegenerateNormV1) {
      return 'invalid_orientation';
    }
  }
  return null;
}

SceneValidationV1 validateSceneV1(Map<dynamic, dynamic> scene) {
  for (final key in kForbiddenSceneTopLevelKeysV1) {
    if (scene.containsKey(key)) {
      return const SceneValidationV1.invalid('forbidden_key');
    }
  }
  if (scene['schemaVersion'] != kSceneSchemaVersionV1) {
    return const SceneValidationV1.invalid('schema_version');
  }
  final rev = scene['revision'];
  if (rev is! int || rev < 0) {
    return const SceneValidationV1.invalid('invalid_revision');
  }

  final listener = scene['listener'];
  if (listener is! Map) {
    return const SceneValidationV1.invalid('missing_listener');
  }
  final listenerErr = _validateObject(listener, expectedType: 'listener');
  if (listenerErr != null) {
    return SceneValidationV1.invalid(listenerErr);
  }

  final sources = scene['sources'];
  final emitters = scene['emitters'];
  if (sources is! List || emitters is! List) {
    return const SceneValidationV1.invalid('invalid_collections');
  }
  if (sources.length > kMaxSourcesV1) {
    return const SceneValidationV1.invalid('too_many_sources');
  }
  if (emitters.length > kMaxEmittersV1) {
    return const SceneValidationV1.invalid('too_many_emitters');
  }

  final ids = <String>{};
  void addId(String id) => ids.add(id);
  addId(listener['id'] as String);

  for (final item in sources) {
    if (item is! Map) {
      return const SceneValidationV1.invalid('invalid_object');
    }
    final err = _validateObject(item, expectedType: 'source');
    if (err != null) return SceneValidationV1.invalid(err);
    final id = item['id'] as String;
    if (!ids.add(id)) {
      return const SceneValidationV1.invalid('duplicate_id');
    }
  }
  for (final item in emitters) {
    if (item is! Map) {
      return const SceneValidationV1.invalid('invalid_object');
    }
    final err = _validateObject(item, expectedType: 'emitter');
    if (err != null) return SceneValidationV1.invalid(err);
    final id = item['id'] as String;
    if (!ids.add(id)) {
      return const SceneValidationV1.invalid('duplicate_id');
    }
  }
  return const SceneValidationV1.valid();
}
